warehouse/accounts/services.py

```diff
@@ -812,3 +814,58 @@ def check_password(self, password, *, tags=None):
         # This service allows *every* password as a non-breached password. It will never
         # tell a user their password isn't good enough.
         return False
+
+
+@implementer(IEmailBreachedService)
+class HaveIBeenPwnedEmailBreachedService:
+    def __init__(
+        self,
+        *,
+        session,
+        api_base="https://haveibeenpwned.com/api/v3/breachedaccount/",
+        api_key=None,
+    ):
+        self._http = session
+        self._api_base = api_base
+        self.api_key = api_key
+
+    @classmethod
+    def create_service(cls, context, request):
+        hibp_api_key = request.registry.settings.get("hibp.api_key")
+        return cls(session=request.http, api_key=hibp_api_key)
+
+    def get_email_breach_count(self, email: str) -> int | str:
+        """
+        Check if an email has been breached, return the number of breaches.
+        See https://haveibeenpwned.com/API/v3#BreachesForAccount
+        """
+
+        # bail early if no api key is set, so we don't send failing requests
+        if not self.api_key:
+            return "!!!"
+
+        try:
+            resp = self._http.get(
+                urllib.parse.urljoin(self._api_base, email),
+                headers={"User-Agent": "PyPI.org", "hibp-api-key": self.api_key},
+            )
+            resp.raise_for_status()
+        except requests.RequestException as exc:
+            # 404 is expected if the email has **not** been breached
+            if exc.response.status_code == http.HTTPStatus.NOT_FOUND:
+                return 0
+            logger.warning("Error contacting HaveIBeenPwned: %r", exc)
+            return "?"
```

Comment:
Feels a bit weird that this returns some semi-cryptic strings instead of integers when things are failing. I think if we ever unset the API key, we would just want this to stop attempting to hit the API and disappear from the UI rather than return `!!!` when it gets here for every email. Similarly for the request exceptions, maybe it'd be better to just return `-1`?

Reply:
Thanks for the critique!
I like that approach - so the return type changes to `int | None`. `None` would then be a signal to the UI to not display the column at all, and -1 is an error condition?